### PR TITLE
V2.2.0

### PR DESCRIPTION
**Pull request by igiloh-pinecone**
New release v.2.20

### CODE REVIEW

**Comment by rajat08** [comment]
@igiloh-pinecone are these changes covered in docs and integration tests?

**Comment by rajat08** [comment]
> @igiloh-pinecone are these changes covered in docs and integration tests?

additionally , are protos in the protos repo up to date?

**Comment by igiloh-pinecone** [comment]
> @igiloh-pinecone are these changes covered in docs and integration tests?

There are UTs in the repo for the recent changes, no new integ tests.  
We have tested the functionality manually against the beta env. We will need to add integ tests as soon as possible.

As for docs - they are documented in the docstrings of the functions themselves, and in the CHANGELOG.

**Comment by rajat08** [comment]
__generally__ we should have all these changes in before we do pypi releases, but it's okay i guess if it was urgent

**Comment by rajat08** [comment]
We have release out in pypi that is from code that's merged yet. We should avoid doing this

**Comment by igiloh-pinecone** [comment]
Anyway I suggest we merge this to `main` now since it's de-facto released anyway.

All the code changes were reviewed and approved in previous individual PRs, and unit tests are passing.  
We can release another minor version with documentation updates in a few days, if we find that anything is missing.

**Review by rajat08 — DISMISSED** [review]

File: pinecone/index.py [hunk 1]
```diff
@@ -151,12 +167,65 @@ def _upsert_batch(self,
 
         args_dict = self._parse_non_empty_args([('namespace', namespace)])
```
Comment (rajat08) [inline pinecone/index.py:168]: nit: this is not needed for upsert, namespace is the only arg

File: pinecone/index.py [hunk 2]
```diff
@@ -169,6 +241,37 @@ def _vector_transform(item: Union[Vector, Tuple]):
             **{k: v for k, v in kwargs.items() if k in _OPENAPI_ENDPOINT_PARAMS}
         )
 
+    def upsert_from_dataframe(self,
+                         df,
+                         namespace: str = None,
+                         batch_size: int = 500,
```
Comment (rajat08) [inline pinecone/index.py:247]: default batch size is too high

File: pinecone/index.py [hunk 3]
```diff
@@ -169,6 +241,37 @@ def _vector_transform(item: Union[Vector, Tuple]):
...
+                         show_progress: bool = True) -> None:
+        """Upserts a dataframe into the index.
+
+        Args:
+            df: A pandas dataframe with the following columns: id, vector, and metadata.
```
Comment (rajat08) [inline pinecone/index.py:252]: why does it not support sparse values? If it does we need to update this doc string

File: pinecone/index.py [hunk 4]
```diff
@@ -169,6 +241,37 @@ def _vector_transform(item: Union[Vector, Tuple]):
...
+            namespace: The namespace to upsert into.
```
Comment (rajat08) [inline pinecone/index.py:253]: Why not have namespace in the schema rather than a explicit arg here

**Comment by rajat08** [comment]
This PR adds numpy, tqdm and pandas dependencies, and I don't see too much value putting it in the client. I am not sure why we are adding them, but I'll defer to product cc @gdj0nes

**Comment by rajat08** [comment]
Nit : In index.py we have delete return `Dict[str, Any] ` , the delete call from the backend sends us nothing but whether the call was accepted, we should stick to that

**Comment by rajat08** [comment]
Nit: the additional args check that we do for `upsert` should be done for all data plane calls

File: pinecone/index.py [hunk 5]
```diff
@@ -151,12 +167,68 @@ def _upsert_batch(self,
 
         args_dict = self._parse_non_empty_args([('namespace', namespace)])
```
Comment (rajat08) [inline pinecone/index.py:168]: nit : I dont think the parse_non_empty_args function is too necessary, we are not doing any extra validation there

**Comment by rajat08** [comment]
I think we have some really nice tests in the repo now. Let's try to make them part of the CI, use tox, fixtures etc.

**Comment by rajat08** [comment]
@igiloh-pinecone  I believe we still need to update grpc version right

File: pinecone/index.py [hunk 6]
```diff
@@ -169,6 +241,37 @@ def _vector_transform(item: Union[Vector, Tuple]):
             **{k: v for k, v in kwargs.items() if k in _OPENAPI_ENDPOINT_PARAMS}
         )
 
+    def upsert_from_dataframe(self,
+                         df,
+                         namespace: str = None,
+                         batch_size: int = 500,
+                         show_progress: bool = True) -> None:
```
Comment (rajat08) [inline pinecone/index.py:248]: We should set this to `false` by default we don't want to print out stuff to users stdout without them expecting it to

File: pinecone/core/grpc/index_grpc.py [hunk 7]
```diff
@@ -274,18 +310,25 @@ def upsert(self,
...
         Examples:
-            >>> index.upsert([('id1', [1.0, 2.0, 3.0], {'key': 'value'}), ('id2', [1.0, 2.0, 3.0])],
-            >>>               namespace='ns1', async_req=True)
+            >>> index.upsert([('id1', [1.0, 2.0, 3.0], {'key': 'value'}),
+                              ('id2', [1.0, 2.0, 3.0])
+                              ],
+                              namespace='ns1', async_req=True)
+            >>> index.upsert([{'id': 'id1', 'values': [1.0, 2.0, 3.0], 'metadata': {'key': 'value'}},
+                              {'id': 'id2',
+                                        'values': [1.0, 2.0, 3.0],
```
Comment (rajat08) [inline pinecone/core/grpc/index_grpc.py:319]: nit : values twice

File: pinecone/index.py (hunk 2 above)
Comment (rajat08) [reply pinecone/index.py:247]: Do we have docs for these interface changes?

**Comment by rajat08** [comment]
My 2 cents: We added tqdm, pandas, and numpy dependencies here that give us little value. I don't see why we need to batch for the users; a user can shoot themselves in the foot here; I lean towards keeping APIs simple unless we do something significantly optimized behind the scenes. Just putting it out there, I would leave the final call to others

File: pinecone/core/grpc/index_grpc.py [hunk 8]
```diff
@@ -359,6 +450,52 @@ def _upsert_batch(self,
...
+            results = [async_result.result() for async_result in tqdm(results, desc="collecting async responses")]
+
+        upserted_count = 0
+        for res in results:
+            upserted_count += res.upserted_count
+
+        return UpsertResponse(upserted_count=upserted_count)
+
+    @staticmethod
+    def _iter_dataframe(df, batch_size):
```
Comment (rajat08) [inline pinecone/core/grpc/index_grpc.py:494]: How does this compare to using `groupBy` and `apply` using an upsert function to batches of rows in the df?

File: pinecone/index.py [hunk 9]
```diff
@@ -169,6 +241,37 @@ def _vector_transform(item: Union[Vector, Tuple]):
...
+            import pandas as pd
+        except ImportError:
+            raise RuntimeError("The `pandas` package is not installed. Please install pandas to use `upsert_from_dataframe()`")
+
+        if not isinstance(df, pd.DataFrame):
+            raise ValueError(f"Only pandas dataframes are supported. Found: {type(df)}")
+
+        upserted_count = 0
+        pbar = tqdm(total=len(df), disable=not show_progress)
+        for i in range(0, len(df), batch_size):
```
Comment (rajat08) [inline pinecone/index.py:267]: this might be overoptimizing but the `itertools.islice` method that we generally use in examples can be more efficient

**Review by miararoy — APPROVED** [review]

**Pull request merged**